sys/net/netutils/util.c

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (C) 2021 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stdlib.h>
+
+#include "net/netif.h"
+#include "net/utils.h"
+#ifdef MODULE_SOCK_DNS
+#include "net/sock/dns.h"
+#endif
+
+/* get the next netif, returns true if there are more */
+static bool _netif_get(netif_t **current_netif)
+{
+    netif_t *netif = *current_netif;
+    netif = netif_iter(netif);
+
+    *current_netif = netif;
+    return netif_iter(netif);
+}
+
+int netutils_parse_hostname(const char *hostname, ipv6_addr_t *addr, netif_t **netif)
```

Comment:
Is there an ip address struct that will handle dual stack? This method signature limits it to IPv6 only (unless v4 mapping in v6 is used)

Reply:
Otherwise maybe add v6 or similar somewhere in the name to make it clearer.

Reply:
Since this comes from `gnrc_icmpv6_echo`, there is no IPv4 support (yet). But good idea, to make the signature protocol independent. @benpicco what do you say? Making it protocol independent and adding a warning in the doc should suffice for now. IPv4 support can be added later.

Reply:
> Otherwise maybe add v6 or similar somewhere in the name to make it clearer.

I would prefer the protocol-independent solution, otherwise we might design the API into a corner we don't want to go into.

Reply:
maybe pass an address family flag to it?

Reply:
How about

```C
int netutils_parse_hostname(const char *hostname, ipv4_addr_t *addr_v4, ipv6_addr_t *addr_v6, netif_t **netif)
```

where either of the `addr` pointers can be NULL?

Reply:
How about making the out parameter for the address an array instead and have an address family parameter (that may or may not be `AF_UNSPEC` for dual stack support later down the road)?

Reply:
this was supposed to be a *convenience* function…

Reply:
Then why do you try make its usage so inconvenient and exotic ;-)?

Reply:
I think this:
> maybe pass an address family to it? And only support inet6 for now.

which I read the same as:
> How about making the out parameter for the address an array instead and have an address family parameter (that may or may not be AF_UNSPEC for dual stack support later down the road)?

make the most sense IMO. 

```suggestion
int netutils_parse_hostname(const char *hostname, void *addr, int family, netif_t **netif)
```

Since there's only support for IPv6 netif's (LWIP doesn't expose `netif`s yet...), it should be enough to `assert(family == AF_INET6)` and keep the current implementation.

Reply:
Note that the original proposal to just add an v6 to the function name has the advantage to allow symbols to be garbage collected when only the v4 or v6 flavor is ever called. Having an explicit family parameter can not as easily be optimized out.

From the user point of view, writing `v6` takes less time than `AF_INET` or `AF_INET6`.

The only reason IMO an additional family parameter is better than separate functions is when `AF_UNSPEC` is supported (as @miri64 suggested). However, I don't think that this easily works with just a byte array, as the caller would still need to now what kind of address is returned. Making the family parameter in/out (`int *family`) would solve this, but that doesn't really sound convenient. The other solution coming to my mind is using something like `[struct sockaddr](https://api.riot-os.org/structsockaddr.html)` to store the address family along with the address. But I personally think that the sockaddr APIs increase complexity and contain quite a few footguns. So I'm personally in favor of simply adding the v6 postfix.

Also note: In both POSIX and standard C functions the out parameters come before the in parameters (except for file descriptors or handles, e.g. see memcpy, strcpy, recvfrom). IMO, C code should generally try to be consistent with this. (I have to admit that getaddrinfo is also breaking with this style, though.)

Reply:
So how about `netutils_get_ipv6()`?

Reply:
Isn't the point of this function that we don't know if it is a plain host name, an IPv4 address or an IPv6 address? So why is it up to the user to tell if this is either?

Reply:
Is there any practically used MCU out there than is indeed using an IPv4/IPv6 dual stack? An `netutils_get_sockaddr()` is IMO an overkill for all use cases but, as it has more overhead and complexity and less type safety and convenience.

One could still implement a `netutils_get_sockaddr()` in addition and let users pick.

Reply:
> Is there any practically used MCU out there than is indeed using an IPv4/IPv6 dual stack?

There are people who use lwIP for dual stack support with RIOT, yes.

> One could still implement a `netutils_get_sockaddr()` in addition and let users pick.

Could generate duplicate code or more overhead, but fine by me (I would rather call it maybe `sock_get_ep(char *hostname)` or something and put it into `sock_util` maybe though ;-)).

Reply:
>  (I would rather call it maybe `sock_get_ep(char *hostname)` or something and put it into `sock_util` maybe though ;-)).

:+1:

Reply:
So `sock_get_ipv6_ep()`?

Reply:
IMHO for just the IPv6 variant and the use-case you are using it in, keeping it in `netutils` should be fine for now. `sock` is not used in those instances and there is a duplication of parts of your code [in `sock_util` anyways already](https://github.com/RIOT-OS/RIOT/blob/0ffd04c132017adf7923eb89594a3a0faab46688/sys/net/sock/sock_util.c#L201-L210)